### PR TITLE
[MIR] Fix vreg flag vector memory leak

### PR DESCRIPTION
**Pull request by optimisan**
A fix-it patch for bec839d8eed9dd13fa7eaffd50b28f8f913de2e2 #110229.

~`SmallVector` will inline 40 values (here) statically on the stack. Although once that is exhausted it'll heap allocate and won't be deallocated.~
No need for a container. This allows 8 flags for a register.

The virtual register flags vector had a memory leak because the vector's memory is not freed.
The `BumpPtrAllocator` handles the deallocation and misses calling the `std::vector<uint8_t> Flags` destructor.

### CODE REVIEW

**Comment by arsenm** [comment]
> The BumpPtrAllocator handles the deallocation and misses calling the std::vector<uint8_t> Flags destructor.

I don't see how this changes that. I would expect this to just move the leak point to cases where the small vector size is exceeded. Whenever VRegInfo is destroyed, the destructor must be called. That's the bug.

But in this case I don't understand why Flags is a vector in the first place. The flags should just be a bitfield, this should be a simple integer field

**Comment by optimisan** [comment]
Was about to comment to say if this is even a right fix, but bitfield would do it.

`VRegInfo *Info = new (Allocator) VRegInfo;` is the allocation site, and there doesn't appear a delete for these, nor did I catch the destructor being called for `VRegInfo`

File: llvm/lib/CodeGen/MIRParser/MIRParser.cpp [hunk 1]
```diff
@@ -703,7 +703,7 @@ bool MIRParserImpl::parseRegisterInfo(PerFunctionMIParsingState &PFS,
         return error(FlagStringValue.SourceRange.Start,
                      Twine("use of undefined register flag '") +
                          FlagStringValue.Value + "'");
-      Info.Flags.push_back(FlagValue);
+      Info.Flags |= FlagValue;
```
Comment (arsenm) [inline llvm/lib/CodeGen/MIRParser/MIRParser.cpp:706]: We should also catch if a flag is set multiple times

**Review by arsenm — APPROVED** [review]

File: llvm/lib/CodeGen/MIRParser/MIRParser.cpp (hunk 1 above)
Comment (optimisan) [reply llvm/lib/CodeGen/MIRParser/MIRParser.cpp:706]: Would a warning be better for it?

Comment (arsenm) [reply llvm/lib/CodeGen/MIRParser/MIRParser.cpp:706]: No reason to be permissive

Comment (optimisan) [reply llvm/lib/CodeGen/MIRParser/MIRParser.cpp:711]: Added the check but can't be tested till relanding the WWM_REG flag patch.

**Review by arsenm — APPROVED** [review]

**Pull request merged**